### ISSUE TITLE
Pretrained model loading

### ISSUE THREAD

**Issue by richardtml** [okankop/Efficient-3DCNNs#24]
Thanks for the repo, it's very useful. I have a couple of questions hoping you can give some insight. 

I would like a use you model for a transfer learning study, from action recognition to another video task. I'am comparing the transferability of the features learned by your model and the ones by TorchVision [ResNet (2+1)D](https://pytorch.org/docs/stable/torchvision/models.html#resnet-2-1-d). I'm using the features just before the fully connected layer. 

Since ShuffleNetV2 is a way smaller, it's normal to expect that ResNet (2+1)D have higher results in the target task. However, the gap between both (63.04 - 70.88 Precision-Recall AUC) is higher than I expected. From my experiments with others models, I expect at most 3 points of difference. So, I have the following questions.

1) Are the following lines correct to load your pretrained model?

```python
# assuming pretrained model is in the current dir
path = 'kinetics_shufflenetv2_1.0x_RGB_16_best.pth'
cp = torch.load(path, map_location=lambda storage, loc: storage)
# remove parallel dataloader prefix
state_dict = {k[7:]: v for k, v in cp['state_dict'].items()}
model = ShuffleNetV2()
model.load_state_dict(state_dict)
```
2) Is it enough normalizing input frames with Kinetics mean/var in [mean](https://github.com/okankop/Efficient-3DCNNs/blob/master/mean.py) module? Or there is another preprocessing step? 

3) Any thoughts on why this could be happening? 

Thank you!

**Comment by okankop** [comment]
Happy to hear that our code and pretrained models are useful for your project!

1. Your code seems correct. You can check which weights are loaded with the following code:
```bash
for k, v in state_dict.items()
    print(k)
```

2. Please check out the validation data loader part for the correct data processing steps. It should be `norm_method = Normalize(opt.mean, [1, 1, 1])`, and opt.mean is the Kinetics mean value.

3.You are using shufflenetv2_1.0x, which performs 9% worse accuracy compared to its 2.0x version on Kinetics dataset. That might be the reason.

**Comment by richardtml** [comment]
Thank you @okankop ! I'll check out your suggestions and I'll be back 😎.

**Comment by ahkarami** [comment]
Dear @okankop & @ahmetgunduz,
Thanks for your great job!!
I have loaded the `kinetics_shufflenet_2.0x_G3_RGB_16_best.pth` pretrained model. Then, for just a simple test of it on a ~ 30 second video (about eating a sandwich) I followed the below steps:
1- read Kinetics-600 class labels according to the **alphabetical order** from this [link](https://gist.github.com/willprice/f19da185c9c5f32847134b87c1960769#file-kinetics_600_labels-csv)
2- read video via OpenCV
3- Since the format (shape) of each video frame is based on **BGR**, so I transformed it to **RGB** at first 
4- Then resize each RGB frame to 112 * 112  # now each frame is numpy.ndarray of (112, 112, 3) 
5- Then apply the below pre-process for each frame:
```python
        # Image pre-processing transforms
        # Kinetics-600 Statistics (mean & std):
        # according to the mean.py file & range [0, 1]:
        self.normalize = transforms.Normalize(
            mean=[0.4338, 0.4045, 0.3776],
            std=[0.1519, 0.1485, 0.1569]            
        )

        self.preprocess = transforms.Compose([
            transforms.ToTensor(),
            self.normalize
        ])
```
6- At this step each frame shape is as: `Normalized Tensor: torch.Size([3, 112, 112])`
7- Then apply above scenario for other frames & collect **16 frames** into a list & after that via the `torch.stack` command change each 16 consecutive frames into a tensor of shape `torch.Size([1, 3, 16, 112, 112]) N, C, T, H, W & N = Batch_Size`
8- Then apply forward pass via your mentioned pretrained model. 
However, with this method, the (aggregated) outputs is not an appropriate one, and the output of the whole video is as follows (Top-3 labels with their associated probabilities):
using a paint roller --> 1.66%, snowkiting --> 1.53%, drawing --> 1.12%
  
However, I have performed exactly the same method using the PyTorch [r2plus1d_18](https://pytorch.org/docs/stable/torchvision/models.html#resnet-2-1-d) model on the same video and got the following good results:
tasting food --> 29.01%, eating burger --> 28.84%  

Would you please kindly guide me in which part I have made a mistake?

Best

**Comment by okankop** [comment]
@ahkarami could you please try using std=[1, 1, 1] for your normalization since it is the one used for validation.

**Comment by ahkarami** [comment]
Dear @okankop,
Thanks for your reply. I have tried it, but the problem is remain. with **std=[1, 1, 1]** the result came as follows:
lunge --> 2.06%, sign language interpreting --> 1.35%, stretching arm --> 1.24%, mountain climber (exercise) --> 1.07%, squat --> 0.87%
which the above produced results aren't correct (and I'm wondering why the produced probabilities are too low (~ 2-1%)). 
If you have another point, I am very grateful to guide me.

**Comment by okankop** [comment]
Hi @ahkarami, models are trained with input in range [0, 255]. In your code, you are normalizing input to [0, 1]. So for your normalization, you should use mean=[110.63666788, 103.16065604, 96.29023126], std=[1, 1, 1]. Moreover, instead of using transforms.ToTensor() (this normalizes input to [0,1]), you can use the ToTensor function in the (spatial_transforms.py)[https://github.com/okankop/Efficient-3DCNNs/blob/master/target_transforms.py]

**Comment by ahkarami** [comment]
Dear @okankop,
Thank you very much for your help. I followed your instructions, but unfortunately again the code did not produce the desired results (at this configuration, at first I changed the format of OpenCV frame from **BGR** to **RGB** via `img_cv = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)`). 
But, when I comment above code (_i.e.,_ using the **BGR** format) then the code produced the desired results :smile: 
When I checked your [code](https://github.com/okankop/Efficient-3DCNNs/blob/355eccbd0495237b562d8336e0ecf2e258db5c04/datasets/kinetics.py#L41), I noticed that apparently you also used the **BGR** format for input images/frames during the training (because you didn't use `cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)`). 
In any case, thank you very much for the useful information you provided to me :pray: .

**Comment by okankop** [comment]
Hi @ahkarami, great to hear that you fixed the issue. I am also happy that you find our work usefull :)
